File: CSharp/OpenTraceability/Mappers/OpenTraceabilityJsonLDMapper.cs
using OpenTraceability.Interfaces;
using OpenTraceability.Models.Events.KDEs;
using OpenTraceability.Models.Events;
using OpenTraceability.Models.Identifiers;
using OpenTraceability.Utility.Attributes;
using OpenTraceability.Utility;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json;
using OpenTraceability.Models.Common;

namespace OpenTraceability.Mappers
{
    public static class OpenTraceabilityJsonLDMapper
    {
        /// <summary>
        /// Converts an object into JSON.
        /// </summary>
        public static JToken ToJson(object value, Dictionary<string, string> namespacesReversed, bool required = false)
        {
            try
            {
                if (value != null)
                {
                    JToken json = new JObject();
                    JToken jpointer = json;

                    Type t = value.GetType();
                    OTMappingTypeInformation typeInfo = OTMappingTypeInformation.GetJsonTypeInfo(t);
                    foreach (var property in typeInfo.Properties.Where(p => p.Version == null || p.Version == EPCISVersion.V2))
                    {
                        object obj = property.Property.GetValue(value);
                        if (obj != null)
                        {
                            JToken jvaluepointer = jpointer;
                            string xchildname = property.Name;

                            if (property.IsQuantityList)
                            {
                                List<EventProduct> products = (List<EventProduct>)obj;
                                products = products.Where(p => p.Quantity != null && p.Type == property.ProductType).ToList();
                                if (products.Count > 0)
                                {
                                    JArray xQuantityList = new JArray();
                                    foreach (var product in products)
                                    {
                                        if (product.EPC != null && product.Quantity != null)
                                        {
                                            JObject xQuantity = new JObject();
                                            xQuantity["epcClass"] = product.EPC.ToString();
                                            xQuantity["quantity"] = product.Quantity.Value;
                                            if (product.Quantity.UoM.UNCode != "EA")
                                            {
                                                xQuantity["uom"] = product.Quantity.UoM.UNCode;
                                            }
                                            xQuantityList.Add(xQuantity);
                                        }
                                    }
                                    jvaluepointer[xchildname] = xQuantityList;
                                }
                            }
                            else if (property.IsEPCList)
                            {
                                List<EventProduct> products = (List<EventProduct>)obj;
                                products = products.Where(p => p.Quantity == null && p.Type == property.ProductType).ToList();
                                if (products.Count > 0 || property.Required)
                                {
                                    JArray xEPCList = new JArray();
                                    foreach (var product in products)
                                    {
                                        if (product.EPC != null)
                                        {
                                            xEPCList.Add(product.EPC.ToString());
                                        }
                                    }
                                    jvaluepointer[xchildname] = xEPCList;
                                }
                            }
                            else if (property.IsArray)
                            {
                                IList list = (IList)obj;
                                JArray xlist = new JArray();

                                if (list.Count > 0 || property.Required == true)
                                {
                                    if (property.IsRepeating && list.Count == 1)
                                    {
                                        JToken jt = WriteObjectToJToken(list[0]);
                                        if (jt != null)
                                        {
                                            jvaluepointer[xchildname] = jt;
                                        }
                                    }
                                    else
                                    {
                                        foreach (var o in list)
                                        {
                                            if (property.IsObject)
                                            {
                                                JToken xchild = ToJson(o, namespacesReversed, property.Required);
                                                if (xchild != null)
                                                {
                                                    xlist.Add(xchild);
                                                }
                                            }
                                            else
                                            {
                                                JToken jt = WriteObjectToJToken(o);
                                                if (jt != null)
                                                {
                                                    xlist.Add(jt);
                                                }
                                            }
                                        }

                                        jvaluepointer[xchildname] = xlist;
                                    }
                                }
                            }
                            else if (property.IsObject)
                            {
                                JToken xchild = ToJson(obj, namespacesReversed, property.Required);
                                if (xchild != null)
                                {
                                    jvaluepointer[xchildname] = xchild;
                                }
                            }
                            else
                            {
                                JToken jt = WriteObjectToJToken(obj);
                                if (jt != null)
                                {
                                    jvaluepointer[xchildname] = jt;
                                }
                            }
                        }
                    }

                    if (typeInfo.ExtensionKDEs != null)
                    {
                        object obj = typeInfo.ExtensionKDEs.GetValue(value);
                        if (obj != null && obj is IList<IEventKDE>)
                        {
                            IList<IEventKDE> kdes = obj as IList<IEventKDE>;
                            if (kdes != null)
                            {
                                foreach (var kde in kdes)
                                {
                                    JToken xchild = kde.GetJson();
                                    if (xchild != null)
                                    {
                                        string name = kde.Name;
                                        if (!string.IsNullOrWhiteSpace(kde.Namespace))
                                        {
                                            if (!namespacesReversed.ContainsKey(kde.Namespace))
                                            {
                                                throw new Exception($"The namespace {kde.Namespace} is not recognized in the EPCIS Document / EPCIS Query Document.");
                                            }
                                            name = namespacesReversed[kde.Namespace] + ":" + name;
                                        }

                                        jpointer[name] = xchild;
                                    }
                                }
                            }
                        }
                    }

                    if (typeInfo.ExtensionAttributes != null)
                    {
                        object obj = typeInfo.ExtensionAttributes.GetValue(value);
                        if (obj != null && obj is IList<IEventKDE>)
                        {
                            IList<IEventKDE> kdes = obj as IList<IEventKDE>;
                            if (kdes != null)
                            {
                                foreach (IEventKDE kde in kdes)
                                {
                                    JToken xchild = kde.GetJson();
                                    if (xchild != null)
                                    {
                                        string name = kde.Name;
                                        if (kde.Namespace != null)
                                        {
                                            if (!namespacesReversed.ContainsKey(kde.Namespace))
                                            {
                                                throw new Exception($"The namespace {kde.Namespace} is not recognized in the EPCIS Document / EPCIS Query Document.");
                                            }
                                            name = namespacesReversed[kde.Namespace] + ":" + name;
                                        }

                                        jpointer[name] = xchild;
                                    }
                                }
                            }
                        }
                    }

                    return json;
                }
                else
                {
                    return null;
                }
            }
            catch (Exception ex)
            {
                Exception e = new Exception($"Failed to parse json. value={value}", ex);
                OTLogger.Error(e);
                throw e;
            }
        }

        /// <summary>
        /// Converts a JSON object into the generic type specified.
        /// </summary>
        public static T FromJson<T>(JToken json, Dictionary<string, string> namespaces)
        {
            T o = (T)FromJson(json, typeof(T), namespaces);
            return o;
        }

        /// <summary>
        /// Converts a JSON object into the type specified.
        /// </summary>
        public static object FromJson(JToken json, Type type, Dictionary<string, string> namespaces)
        {
            object value = Activator.CreateInstance(type) ?? throw new Exception("Failed to create instance of type " + type.FullName);

            try
            {
                OTMappingTypeInformation typeInfo = OTMappingTypeInformation.GetJsonTypeInfo(type);

                List<IEventKDE> extensionKDEs = null;
                List<IEventKDE> extensionAttributes = null;

                if (typeInfo.ExtensionAttributes != null)
                {
                    extensionAttributes = new List<IEventKDE>();
                }

                if (typeInfo.ExtensionKDEs != null)
                {
                    extensionKDEs = new List<IEventKDE>();
                }

                OTMappingTypeInformationProperty mappingProp;

                JObject jobj = json as JObject;
                if (jobj != null)
                {
                    foreach (JProperty jprop in jobj.Properties())
                    {
                        mappingProp = typeInfo[jprop.Name];

                        if (mappingProp != null && mappingProp.Property.SetMethod == null)
                        {
                            continue;
                        }

                        JToken jchild = jobj[jprop.Name];
                        if (jchild != null)
                        {
                            if (mappingProp != null)
                            {
                                ReadPropertyMapping(mappingProp, jchild, value, namespaces);
                            }
                            else if (extensionKDEs != null)
                            {
                                IEventKDE kde = ReadKDE(jprop.Name, jchild, namespaces);
                                extensionKDEs.Add(kde);
                            }
                            else if (extensionAttributes != null)
                            {
                                IEventKDE kde = ReadKDE(jprop.Name, jchild, namespaces);
                                extensionAttributes.Add(kde);
                            }
                        }
                    }
                }

                if (typeInfo.ExtensionAttributes != null)
                {
                    typeInfo.ExtensionAttributes.SetValue(value, extensionAttributes);
                }

                if (typeInfo.ExtensionKDEs != null)
                {
                    typeInfo.ExtensionKDEs.SetValue(value, extensionKDEs);
                }
            }
            catch (Exception ex)
            {
                OTLogger.Error(ex);
                throw;
            }

            return value;
        }

        private static JToken WriteObjectToJToken(object obj)
        {
            if (obj == null)
            {
                return null;
            }
            else if (obj is List<LanguageString>)
            {
                string json = JsonConvert.SerializeObject(obj);
                return JArray.Parse(json);
            }
            else if (obj is DateTimeOffset)
            {
                DateTimeOffset dt = (DateTimeOffset)obj;
                return dt.ToString("O");
            }
            else if (obj is UOM)
            {
                UOM uom = (UOM)obj;
                return uom.UNCode;
            }
            else if (obj is double)
            {
                return JToken.FromObject(obj);
            }
            else if (obj is bool)
            {
                return JToken.FromObject(obj);
            }
            else if (obj is Country)
            {
                Country b = (Country)obj;
                return b.Abbreviation;
            }
            else if (obj is TimeSpan)
            {
                TimeSpan timespan = (TimeSpan)obj;
                if (timespan.Ticks < 0)
                {
                    return "-" + timespan.Negate().TotalHours.ToString("#00") + ":" + timespan.Minutes.ToString("00");
                }
                else
                {
                    return "+" + timespan.TotalHours.ToString("#00") + ":" + timespan.Minutes.ToString("00");
                }
            }
            else
            {
                return obj.ToString() ?? string.Empty;
            }
        }

        private static void ReadPropertyMapping(OTMappingTypeInformationProperty mappingProp, JToken json, object value, Dictionary<string, string> namespaces)
        {
            if (mappingProp.IsQuantityList)
            {
                IEvent e = (IEvent)value;
                JArray jQuantityList = json as JArray;
                if (jQuantityList != null)
                {
                    foreach (JObject jQuantity in jQuantityList)
                    {
                        EPC epc = new EPC(jQuantity["epcClass"]?.Value<string>() ?? string.Empty);

                        EventProduct product = new EventProduct(epc);
                        product.Type = mappingProp.ProductType;

                        double quantity = jQuantity.Value<double>("quantity");
                        string uom = jQuantity.Value<string>("uom") ?? "EA";
                        product.Quantity = new Measurement(quantity, uom);

                        e.AddProduct(product);
                    }
                }
            }
            else if (mappingProp.IsEPCList)
            {
                IEvent e = (IEvent)value;
                JArray jEPCList = json as JArray;
                if (jEPCList != null)
                {
                    foreach (JToken jEPC in jEPCList)
                    {
                        EPC epc = new EPC(jEPC.ToString());
                        EventProduct product = new EventProduct(epc);
                        product.Type = mappingProp.ProductType;
                        e.AddProduct(product);
                    }
                }
            }
            else if (mappingProp.IsArray)
            {
                IList list = mappingProp.Property.GetValue(value) as IList;
                if (list == null)
                {
                    list = (IList)(Activator.CreateInstance(mappingProp.Property.PropertyType)
                        ?? throw new Exception("Failed to create instance of " + mappingProp.Property.PropertyType.FullName));

                    mappingProp.Property.SetValue(value, list);
                }

                Type itemType = list.GetType().GenericTypeArguments[0];

                if (mappingProp.IsRepeating && !(json is JArray))
                {
                    string v = json.ToString();
                    if (!string.IsNullOrWhiteSpace(v))
                    {
                        object o = ReadObjectFromString(v, itemType);
                        list.Add(o);
                    }
                }
                else
                {
                    JArray jArr = json as JArray;
                    if (jArr != null)
                    {
                        foreach (JToken j in jArr)
                        {
                            if (mappingProp.IsObject)
                            {
                                object o = FromJson(j, itemType, namespaces);
                                list.Add(o);
                            }
                            else
                            {
                                object o = ReadObjectFromString(j.ToString(), itemType);
                                list.Add(o);
                            }
                        }
                    }
                }
            }
            else if (mappingProp.IsObject)
            {
                object o = FromJson(json, mappingProp.Property.PropertyType, namespaces);
                mappingProp.Property.SetValue(value, o);
            }
            else if (mappingProp.Property.PropertyType == typeof(List<LanguageString>))
            {
                List<LanguageString> languageStrings = JsonConvert.DeserializeObject<List<LanguageString>>(json.ToString());
                if (languageStrings != null)
                {
                    mappingProp.Property.SetValue(value, languageStrings);
                }
            }
            else
            {
                string v = json.ToString();
                if (!string.IsNullOrWhiteSpace(v))
                {
                    object o = ReadObjectFromString(v, mappingProp.Property.PropertyType);
                    mappingProp.Property.SetValue(value, o);
                }
            }
        }

        private static object ReadObjectFromString(string value, Type t)
        {
            try
            {
                if (t == typeof(DateTimeOffset) || t == typeof(DateTimeOffset?))
                {
                    DateTimeOffset dt = value.TryConvertToDateTimeOffset() ?? throw new Exception("Failed to convert string to datetimeoffset where value = " + value);
                    return dt;
                }
                else if (t == typeof(UOM))
                {
                    UOM uom = UOM.LookUpFromUNCode(value);
                    return uom;
                }
                else if (t == typeof(bool) || t == typeof(bool?))
                {
                    bool v = bool.Parse(value);
                    return v;
                }
                else if (t == typeof(double) || t == typeof(double?))
                {
                    double v = double.Parse(value);
                    return v;
                }
                else if (t == typeof(Uri))
                {
                    Uri v = new Uri(value);
                    return v;
                }
                else if (t == typeof(TimeSpan) || t == typeof(TimeSpan?))
                {
                    if (value.StartsWith("+")) value = value.Substring(1);
                    TimeSpan ts = TimeSpan.Parse(value);
                    return ts;
                }
                else if (t == typeof(EventAction) || t == typeof(EventAction?))
                {
                    EventAction action = (EventAction)Enum.Parse(typeof(EventAction), value);
                    return action;
                }
                else if (t == typeof(PGLN))
                {
                    PGLN pgln = new PGLN(value);
                    return pgln;
                }
                else if (t == typeof(GLN))
                {
                    GLN gln = new GLN(value);
                    return gln;
                }
                else if (t == typeof(GTIN))
                {
                    GTIN gtin = new GTIN(value);
                    return gtin;
                }
                else if (t == typeof(EPC))
                {
                    EPC epc = new EPC(value);
                    return epc;
                }
                else if (t == typeof(Country))
                {
                    Country c = Countries.Parse(value);
                    return c;
                }
                else
                {
                    return value;
                }
            }
            catch (Exception ex)
            {
                Exception e = new Exception($"Failed to convert string into object. value={value} and t={t}", ex);
                OTLogger.Error(e);
                throw e;
            }
        }

        private static IEventKDE ReadKDE(string name, JToken json, Dictionary<string, string> namespaces)
        {
            IEventKDE kde = null;

            //if not, check if it is a simple value or an object
            if (kde == null)
            {
                string ns = string.Empty;
                if (name.Contains(":"))
                {
                    ns = name.Split(':').First();
                    name = name.Split(':').Last();

                    if (!namespaces.ContainsKey(ns))
                    {
                        throw new Exception("The KDE has a namespace prefix, but there is no such namespace in the dictionary. " + ns);
                    }
                    ns = namespaces[ns];
                }

                if (json is JObject || json is JArray)
                {
                    kde = new EventKDEObject(ns, name);
                }
                //else if simple value, then we will consume it as a string
                else
                {
                    kde = new EventKDEString(ns, name);
                }
            }

            if (kde != null)
            {
                kde.SetFromJson(json);
            }
            else
            {
                throw new Exception("Failed to initialize KDE from JSON = " + json.ToString());
            }

            return kde;
        }
    }
}

File: JDK11/opentraceability/src/main/java/opentraceability/mappers/opentraceabilityjsonldmapper.java
package opentraceability.mappers;

import opentraceability.interfaces.*;
import opentraceability.models.events.kdes.*;
import opentraceability.models.events.*;
import opentraceability.models.identifiers.*;
import opentraceability.utility.*;
import opentraceability.models.common.*;
import opentraceability.*;
import org.json.JSONArray;
import org.json.JSONObject;

import java.lang.reflect.Type;
import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

public final class OpenTraceabilityJsonLDMapper
{
	/** 
	 Converts an object into JSON.
	*/

	public static Object ToJson(Object value, java.util.HashMap<String, String> namespacesReversed)
	{
		return ToJson(value, namespacesReversed, false);
	}

//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: public static System.Nullable<Object> ToJson(object? value, Dictionary<string, string> namespacesReversed, bool required = false)
//C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
	public static Object ToJson(Object value, HashMap<String, String> namespacesReversed, boolean required) throws IllegalAccessException {
		if (value != null)
		{
			JSONObject json = new JSONObject();
			JSONObject jpointer = json;

			Type t = value.getClass();
			OTMappingTypeInformation typeInfo = OTMappingTypeInformation.getJsonTypeInfo(t);
			for (var property : typeInfo.properties.stream().filter(p -> p.version == null || p.version == EPCISVersion.V2).collect(Collectors.toList()))
			{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: object? obj = property.Property.get(value);
				Object obj = property.field.get(value);
				if (obj != null)
				{
					Object jvaluepointer = jpointer;
					String xchildname = property.name;

					if (property.isQuantityList)
					{
						ArrayList<EventProduct> products = (ArrayList<EventProduct>)obj;
						products = products.stream().filter(p -> p.Quantity != null && p.Type == property.ProductType).collect(Collectors.toList());
						if (!products.isEmpty())
						{
							JSONArray xQuantityList = new JSONArray();
							for (var product : products)
							{
								if (product.EPC != null && product.Quantity != null)
								{
									JSONObject jQuantity = new JSONObject();
									jQuantity["epcClass"] = product.EPC.toString();
									jQuantity["quantity"] = product.Quantity.getValue();
									if (!Objects.equals(product.Quantity.getUoM().getUNCode(), "EA"))
									{
										jQuantity["uom"] = product.Quantity.getUoM().getUNCode();
									}
									xQuantityList.Add(xQuantity);
								}
							}
							jvaluepointer[xchildname] = xQuantityList;
						}
					}
					else if (property.IsEPCList)
					{
						ArrayList<EventProduct> products = (ArrayList<EventProduct>)obj;
						products = products.stream().filter(p -> p.Quantity == null && p.Type == property.ProductType).collect(Collectors.toList());
						if (!products.isEmpty() || property.Required)
						{
							JSONArray xEPCList = new JSONArray();
							for (var product : products)
							{
								if (product.EPC != null)
								{
									xEPCList.Add(product.EPC.toString());
								}
							}
							jvaluepointer[xchildname] = xEPCList;
						}
					}
					else if (property.IsArray)
					{
						List list = (List)obj;
						JSONArray xlist = new JSONArray();

						if (!list.isEmpty() || property.Required == true)
						{
							if (property.IsRepeating && list.size() == 1)
							{
								Object jt = WriteObjectToJToken(list.get(0));
								if (jt != null)
								{
									jvaluepointer[xchildname] = jt;
								}
							}
							else
							{
								for (var o : list)
								{
									if (property.IsObject)
									{
										Object xchild = ToJson(o, namespacesReversed, property.Required);
										if (xchild != null)
										{
											xlist.Add(xchild);
										}
									}
									else
									{
										Object jt = WriteObjectToJToken(o);
										if (jt != null)
										{
											xlist.Add(jt);
										}
									}
								}

								jvaluepointer[xchildname] = xlist;
							}
						}
					}
					else if (property.IsObject)
					{
						Object xchild = ToJson(obj, namespacesReversed, property.Required);
						if (xchild != null)
						{
							jvaluepointer[xchildname] = xchild;
						}
					}
					else
					{
						Object jt = WriteObjectToJToken(obj);
						if (jt != null)
						{
							jvaluepointer[xchildname] = jt;
						}
					}
				}
			}

			if (typeInfo.extensionKDEs != null)
			{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: object? obj = typeInfo.ExtensionKDEs.get(value);
				Object obj = typeInfo.extensionKDEs.get(value);
				if (obj != null && obj instanceof List<IEventKDE>)
				{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: IList<IEventKDE>? kdes = obj instanceof java.util.List<IEventKDE> ? (java.util.List<IEventKDE>)obj : null;
					List<IEventKDE> kdes = obj instanceof List<IEventKDE> ? (List<IEventKDE>)obj : null;
					if (kdes != null)
					{
						for (var kde : kdes)
						{
							Object xchild = kde.getJson();
							if (xchild != null)
							{
								String name = kde.getName();
								if (kde.namespaces != null)
								{
									if (!namespacesReversed.containsKey(kde.getNamespace()))
									{
										throw new RuntimeException(String.format("The namespace %1$s is not recognized in the EPCIS Document / EPCIS Query Document.", kde.getNamespace()));
									}
									name = namespacesReversed.get(kde.getNamespace()) + ":" + name;
								}

								jpointer[name] = xchild;
							}
						}
					}
				}
			}

			if (typeInfo.getExtensionAttributes() != null)
			{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: object? obj = typeInfo.ExtensionAttributes.get(value);
				Object obj = typeInfo.getExtensionAttributes().get(value);
				if (obj != null && obj instanceof List<IEventKDE>)
				{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: IList<IEventKDE>? kdes = obj instanceof java.util.List<IEventKDE> ? (java.util.List<IEventKDE>)obj : null;
					List<IEventKDE> kdes = obj instanceof List<IEventKDE> ? (List<IEventKDE>)obj : null;
					if (kdes != null)
					{
						for (IEventKDE kde : kdes)
						{
							Object xchild = kde.GetJson();
							if (xchild != null)
							{
								String name = kde.getName();
								if (kde.namespaces != null)
								{
									if (!namespacesReversed.containsKey(kde.getNamespace()))
									{
										throw new RuntimeException(String.format("The namespace %1$s is not recognized in the EPCIS Document / EPCIS Query Document.", kde.getNamespace()));
									}
									name = namespacesReversed.get(kde.getNamespace()) + ":" + name;
								}

								jpointer[name] = xchild;
							}
						}
					}
				}
			}

			return json;
		}
		else
		{
			return null;
		}
	}

	/** 
	 Converts a JSON object into the generic type specified.
	*/
	public static <T> T FromJson(Object json, HashMap<String, String> namespaces)
	{
		T o = (T)FromJson(json, T.class, namespaces);
		return o;
	}

	/** 
	 Converts a JSON object into the type specified.
	*/
	public static Object FromJson(Object json, Type type, HashMap<String, String> namespaces) throws Exception {
//C# TO JAVA CONVERTER TASK: Throw expressions are not converted by C# to Java Converter:
//ORIGINAL LINE: object value = Activator.CreateInstance(type) ?? throw new Exception("Failed to create instance of type " + type.FullName);
		Object value = ReflectionUtility.constructType(type);

		OTMappingTypeInformation typeInfo = OTMappingTypeInformation.getJsonTypeInfo(type);

		ArrayList<IEventKDE> extensionKDEs = null;
		ArrayList<IEventKDE> extensionAttributes = null;

		if (typeInfo.extensionAttributes != null)
		{
			extensionAttributes = new ArrayList<IEventKDE>();
		}

		if (typeInfo.extensionKDEs != null)
		{
			extensionKDEs = new ArrayList<IEventKDE>();
		}

//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: OTMappingTypeInformationProperty? mappingProp;
		OTMappingTypeInformationProperty mappingProp;

		JSONObject jobj = json instanceof JSONObject ? (JSONObject)json : null;
		if (jobj != null)
		{
			for (String jprop : jobj.keySet())
			{
				mappingProp = typeInfo.get(jprop);

				if (jobj.has(jprop))
				{
					Object jchild = jobj.get(jprop);
					if (mappingProp != null)
					{
						ReadPropertyMapping(mappingProp, jchild, value, namespaces);
					}
					else if (extensionKDEs != null)
					{
						IEventKDE kde = ReadKDE(jprop, jchild, namespaces);
						extensionKDEs.add(kde);
					}
					else if (extensionAttributes != null)
					{
						IEventKDE kde = ReadKDE(jprop, jchild, namespaces);
						extensionAttributes.add(kde);
					}
				}
			}
		}

		if (typeInfo.extensionAttributes != null)
		{
			typeInfo.extensionAttributes.set(value, extensionAttributes);
		}

		if (typeInfo.extensionKDEs != null)
		{
			typeInfo.extensionKDEs.set(value, extensionKDEs);
		}

		return value;
	}

//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: private static System.Nullable<Object> WriteObjectToJToken(object? obj)
	private static Object WriteObjectToJToken(Object obj)
	{
		if (obj == null)
		{
			return null;
		}
		else if (obj instanceof ArrayList<LanguageString>)
		{
			String json = JsonConvert.SerializeObject(obj);
			return JSONArray.Parse(json);
		}
		else if (obj instanceof OffsetDateTime)
		{
			OffsetDateTime dt = (OffsetDateTime)obj;
			return dt.format(DateTimeFormatter.ISO_DATE_TIME);
		}
		else if (obj instanceof UOM)
		{
			UOM uom = (UOM)obj;
			return uom.UNCode;
		}
		else if (obj instanceof Double)
		{
			return obj;
		}
		else if (obj instanceof Boolean)
		{
			return obj;
		}
		else if (obj instanceof Country)
		{
			Country b = (Country)obj;
			return b.abbreviation;
		}
		else if (obj instanceof Duration)
		{
			Duration timespan = (Duration)obj;
			String timeStr = (new Double(timespan.toHoursPart())).toString() + ":" + (new Integer(timespan.Minutes)).toString("00");
			if (timespan.isNegative())
			{
				return "-" + timeStr;
			}
			else
			{
				return "-" + timeStr;
			}
		}
		else
		{
			return obj.toString() != null ? obj.toString() : "";
		}
	}

	private static void ReadPropertyMapping(OTMappingTypeInformationProperty mappingProp, Object json, Object value, HashMap<String, String> namespaces) throws Exception {
		if (mappingProp.isQuantityList)
		{
			IEvent e = (IEvent)value;
			JSONArray jQuantityList = json instanceof JSONArray ? (JSONArray)json : null;
			if (jQuantityList != null)
			{
				for (Object o : jQuantityList)
				{
					if (o instanceof JSONObject)
					{
						JSONObject jQuantity = (JSONObject) o;
						EPC epc = new EPC((jQuantity.getString("epcClass"));

						EventProduct product = new EventProduct(epc);
						product.Type = mappingProp.productType;
						product.Quantity = JSONExtensions.readMeasurement(jQuantity);

						e.addProduct(product);
					}
				}
			}
		}
		else if (mappingProp.isEPCList)
		{
			IEvent e = (IEvent)value;
			JSONArray jEPCList = json instanceof JSONArray ? (JSONArray)json : null;
			if (jEPCList != null)
			{
				for (Object jEPC : jEPCList)
				{
					EPC epc = new EPC(jEPC.toString());
					EventProduct product = new EventProduct(epc);
					product.Type = mappingProp.productType;
					e.addProduct(product);
				}
			}
		}
		else if (mappingProp.isArray)
		{
			List list = (List)ReflectionUtility.constructType(mappingProp.field.getType());
			Type itemType = ReflectionUtility.getItemType(mappingProp.field.getType());

			if (mappingProp.isRepeating && !(json instanceof JSONArray))
			{
				String v = json.toString();
				if (!(v == null || v.isBlank()))
				{
					Object o = ReadObjectFromString(v, itemType);
					list.add(o);
				}
			}
			else
			{
				JSONArray jArr = json instanceof JSONArray ? (JSONArray)json : null;
				if (jArr != null)
				{
					for (Object j : jArr)
					{
						if (mappingProp.isObject)
						{
							Object o = FromJson(j, itemType, namespaces);
							list.add(o);
						}
						else
						{
							Object o = ReadObjectFromString(j.toString(), itemType);
							list.add(o);
						}
					}
				}
			}
		}
		else if (mappingProp.isObject)
		{
			Object o = FromJson(json, mappingProp.field.getType(), namespaces);
			mappingProp.field.set(value, o);
		}
		else if (mappingProp.field.getType() == ArrayList<LanguageString>.class)
		{
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: List<LanguageString>? languageStrings = JsonConvert.DeserializeObject<List<LanguageString>>(json.ToString());
			ArrayList<LanguageString> languageStrings = JsonConvert.<ArrayList<LanguageString>>DeserializeObject(json.toString());
			if (languageStrings != null)
			{
				mappingProp.field.set(value, languageStrings);
			}
		}
		else
		{
			String v = json.toString();
			if (!(v == null || v.isBlank()))
			{
				Object o = ReadObjectFromString(v, mappingProp.field.getType());
				mappingProp.field.set(value, o);
			}
		}
	}

	private static Object ReadObjectFromString(String value, Type t)
	{
		return ReflectionUtility.parseFromString(t, value);
	}

	private static IEventKDE ReadKDE(String name, Object value, HashMap<String, String> namespaces) throws Exception {
//C# TO JAVA CONVERTER WARNING: Nullable reference types have no equivalent in Java:
//ORIGINAL LINE: IEventKDE? kde = null;
		IEventKDE kde = null;

		//if not, check if it is a simple value or an object
		if (kde == null)
		{
			String ns = "";
			if (name.contains(":"))
			{
				ns = StringExtensions.First(name.split(java.util.regex.Pattern.quote(":"), -1));
				name = StringExtensions.Last(name.split(java.util.regex.Pattern.quote(":"), -1));

				if (!namespaces.containsKey(ns))
				{
					throw new RuntimeException("The KDE has a namespace prefix, but there is no such namespace in the dictionary. " + ns);
				}
				ns = namespaces.get(ns);
			}

			if (value instanceof JSONObject || value instanceof JSONArray)
			{
				kde = new EventKDEObject(ns, name);
			}
			//else if simple value, then we will consume it as a string
			else
			{
				kde = new EventKDEString(ns, name);
			}
		}

		if (kde != null)
		{
			kde.setFromJson(value);
		}
		else
		{
			throw new RuntimeException("Failed to initialize KDE from JSON = " + value.toString());
		}

		return kde;
	}
}
